Gather comments from headers

CGMossa opened the pull request:
Adding these lines to the `clang`-call enables getting some comments.

This is not much, but it is *something*. Maybe it will be more useful later

CGMossa commented:
/bindings

CGMossa commented:
This is ready for review. I just don't know who cares enough to approve / disapprove of this.

multimeric commented:
Ooh, I like this. Two points of feedback. One, I note that almost all of the doc comments have a leading space that at least visually looks annoying. If rustdoc is not stripping out this space then it's probably annoying enough to fix. The other thing is just that I'd like to see an extendr build run using these bindings just to run it through the tests and linting etc. Otherwise it looks good and I think it would be valuable for sure.

CGMossa commented:
There is an option to use `rustfmt` for the generated binding. Issue is it is completely broken for these headers.
Instead, I opted to use an instrospection thing, called `ParseCallbacks`, that allows one to edit what bindgen does in a trait-way.
I implemented a parse_comment-method;

I have investigated this to see if we had something for #139, but this parse-callback just provides you with this information as it goes through, the name of the macro, and the tokens of the macro, and you ought to do something else with it. There is no return nothing. But that's an aside.

CGMossa commented:
/bindings

CGMossa commented:
For the other point, see https://github.com/extendr/extendr/pull/497, once it finishes whatever it needs to do.

CGMossa commented:
/bindgen

CGMossa commented:
Okay, so the answer is, that this `[redacted].clang_arg("-fretain-comments-from-system-headers")` is the problem.
I don't know why, and I cannot check this with a Mac machine.
I guess it could be a todo for someone with a Mac.

CGMossa commented:
/bindgen

CGMossa commented:
/bindings

CGMossa commented:
Alright; I'm re-running everything, but it already worked out, and I've addressed comments from you @multimeric.
PS: I wrote this message on the wrong PR.
But this one is ready for round two.

multimeric reviewed (approved):
Looks great!

yutannihilation commented on build.rs:
```diff
@@ -508,18 +509,24 @@ fn generate_bindings(r_paths: &InstallationPaths, version_info: &RVersionInfo) {
+        .generate_comments(true)
+        .parse_callbacks(Box::new(RCallbacks))
+        .clang_arg("-fparse-all-comments")
+        // .clang_arg("-fretain-comments-from-system-headers")
```
Do we need this line?

```suggestion
```

yutannihilation reviewed (approved):
Looks good to me!

CGMossa commented:
No, but https://github.com/extendr/libR-sys/pull/141#issuecomment-1448804819


On Wed, 1 Mar 2023, 00.50 Hiroaki Yutani, ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> Looks good to me!
> ------------------------------
>
> In build.rs
> <https://github.com/extendr/libR-sys/pull/141#discussion_r1120942606>:
>
> > @@ -508,18 +509,24 @@ fn generate_bindings(r_paths: &InstallationPaths, version_info: &RVersionInfo) {
>
>
>
>      // Finish the builder and generate the bindings.
>
>      let bindings = bindgen_builder
>
> +        .generate_comments(true)
>
> +        .parse_callbacks(Box::new(RCallbacks))
>
> +        .clang_arg("-fparse-all-comments")
>
> +        // .clang_arg("-fretain-comments-from-system-headers")
>
>
> Do we need this line?
> ⬇️ Suggested change
>
> -        // .clang_arg("-fretain-comments-from-system-headers")
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/extendr/libR-sys/pull/141#pullrequestreview-1318839610>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDVSGIHPGL6224JMPRCELWZ2FM7ANCNFSM6AAAAAAVJ47XZM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

CGMossa commented:
It didn't do anything on the windows part.
But it clearly affects the other platforms. So there could be something
there...

I'll remove it (or either of you can remove the line) and then let's merge
this.
It is not clean to have commented lines or code anyways.

On Wed, 1 Mar 2023, 04.59 Michael Milton, ***@***.***> wrote:

> So you removed it because it was causing the build to fail on Mac? What
> value did it provide? I'm sure it's fine to leave for later.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/extendr/libR-sys/pull/141#issuecomment-1449296658>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDVSHEENYT4BWXLJMGPZLWZ3CTNANCNFSM6AAAAAAVJ47XZM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

multimeric commented:
Urgh I just remembered the changelog. Please add a PR updating it too.

yutannihilation commented:
Thanks! Putting it behind `#[cfg(not(target_os = "macos"))]` can be an option, but I'm fine with either.
